lib/receiver.ts

```diff
@@ -0,0 +1,289 @@
+   */
+  isSessionClosed(): boolean {
+    return this._session.isClosed();
+  }
+
+  /**
+   * Determines whether both local and remote endpoint for just the session itself are closed.
+   * Within the "session_close" event handler, if this method returns `false` it means that
+   * the local end is still open. It can be useful to determine whether the close
+   * was initiated locally under such circumstances.
+   *
+   * @returns {boolean} `true` - closed, `false` otherwise.
+   */
+  isSessionItselfClosed(): boolean {
+    return this._session.isItselfClosed();
+  }
+
+  /**
+   * Removes the receiver and it's underlying session from the internal map.
+   * @returns {void} void
+   */
+  remove(): void {
+    if (this._receiver) {
+      this._receiver.remove();
+    }
+    if (this._session) {
+      this._session.remove();
+    }
+  }
+
+  /**
+   * Closes the amqp receiver.
+   * @return {Promise<void>} Promise<void>
+   * - **Resolves** the promise when rhea emits the "receiver_close" event.
+   * - **Rejects** the promise with an AmqpError when rhea emits the
+   * "receiver_error" event while trying to close an amqp receiver.
+   */
+  close(): Promise<void> {
+    const receiverClose = new Promise<void>((resolve, reject) => {
+      log.error("[%s] The receiver is open ? -> %s", this.connection.id, this.isOpen());
+      if (this.isOpen()) {
+        let onError: Func<rhea.EventContext, void>;
+        let onClose: Func<rhea.EventContext, void>;
+        let waitTimer: any;
+
+        const removeListeners = () => {
+          clearTimeout(waitTimer);
+          this._receiver.removeListener(ReceiverEvents.receiverError, onError);
+          this._receiver.removeListener(ReceiverEvents.receiverClose, onClose);
+        };
+
+        onClose = (context: rhea.EventContext) => {
+          removeListeners();
+          process.nextTick(() => {
+            log.receiver("[%s] Resolving the promise as the amqp receiver has been closed.",
+              this.connection.id);
+            resolve();
+          });
+        };
+
+        onError = (context: rhea.EventContext) => {
+          removeListeners();
+          log.error("[%s] Error occurred while closing amqp receiver. %O",
+            this.connection.id, context.session!.error);
+          reject(context.session!.error);
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const msg: string = `Unable to close the amqp receiver ${this.name} due to operation timeout.`;
+          log.error("[%s] %s", this.connection.id, msg);
+          reject(new Error(msg));
+        };
+
+        this._receiver.once(ReceiverEvents.receiverClose, onClose);
+        this._receiver.once(ReceiverEvents.receiverError, onError);
+        waitTimer = setTimeout(actionAfterTimeout, defaultOperationTimeoutInSeconds * 1000);
+        this._receiver.close();
+      } else {
+        resolve();
+      }
+    });
+
+    return receiverClose.then(() => {
```

Comment:
nit: i'd make the outside function `async` and then do:

```ts
await new Promise((resolve, reject) => { /* receiver close logic */ });
return this._session.close();
```

this keeps the patterns more consistent such that anything outside of a `new Promise()` is done with async/await